torch/distributed/fsdp/_common_utils.py

```diff
@@ -4,25 +4,38 @@
 
 import traceback
 from enum import auto, Enum
-from typing import Any, Callable, Dict, List, no_type_check, Set, Union
+from typing import Callable, Dict, Generator, List, no_type_check, Optional, Set
 
 import torch
 import torch.distributed.fsdp.flat_param as flat_param_file
 import torch.nn as nn
+from torch.distributed._composable_state import _get_module_state, _State
 from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import (
     _CHECKPOINT_PREFIX,
 )
 
+from .api import FullStateDictConfig, StateDictConfig, StateDictType
+
 FSDP_WRAPPED_MODULE = "_fsdp_wrapped_module"
 FSDP_PREFIX = FSDP_WRAPPED_MODULE + "."
 FSDP_FLATTENED = "_fsdp_flattened"
 
 
-# We leverage Python's dynamic attribute definition to unify the state
-# management for the wrapper and non-wrapper approaches. The `Any` represents
-# the `_State` object in _composable/contract.py, but we do not import it to
-# avoid circular imports.
-_FSDPState = Union[nn.Module, Any]
+class _FSDPState(_State):
+    def __init__(self) -> None:
+        # Move all the attributes to this class to enable typing for FSDP/fully_shard.
```

Comment:
This is awesome. I think it would be great to get type checking formalized for FSDP. I will need to follow-up to see if there are any cases that require some redefinition of data structures.